velox/functions/prestosql/tests/ArrayShuffleTest.cpp

```diff
@@ -0,0 +1,316 @@
+ * limitations under the License.
+ */
+
+#include <folly/container/F14Set.h>
+#include <algorithm>
+#include <numeric>
+#include <optional>
+#include <random>
+#include <sstream>
+
+#include "velox/expression/VectorReaders.h"
+#include "velox/functions/prestosql/tests/utils/FunctionBaseTest.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::test;
+using namespace facebook::velox::functions::test;
+
+namespace {
+template <typename T>
+std::string printArray(const std::vector<std::optional<T>>& input) {
+  std::stringstream out;
+  out << "[";
+  std::for_each(input.begin(), input.end(), [&out](const auto& e) {
+    if (e.has_value()) {
+      out << e.value();
+    } else {
+      out << "NULL";
+    }
+    out << ", ";
+  });
+  out << "]";
+  return out.str();
+}
+
+template <typename T>
+std::string printArray(
+    const std::vector<std::optional<std::vector<std::optional<T>>>>& input) {
+  std::stringstream out;
+  out << "[";
+  std::for_each(input.begin(), input.end(), [&out](const auto& e) {
+    if (e.has_value()) {
+      out << printArray(e.value());
+    } else {
+      out << "NULL";
+    }
+    out << ", ";
+  });
+  out << "]";
+  return out.str();
+}
+
+template <typename T>
+std::string printArray(const std::vector<T>& input) {
+  std::stringstream out;
+  out << "[";
+  std::for_each(
+      input.begin(), input.end(), [&out](const auto& e) { out << e << ", "; });
+  out << "]";
+  return out.str();
+}
+
+template <typename T, bool returnsOptionalValues>
+exec::ArrayView<returnsOptionalValues, T> read(
+    exec::VectorReader<Array<T>>& reader,
+    size_t offset) {
+  if constexpr (returnsOptionalValues) {
+    return reader[offset];
+  } else {
+    return reader.readNullFree(offset);
+  }
+}
+} // namespace
+
+namespace {
+class ArrayShuffleTest : public FunctionBaseTest {
+ protected:
+  constexpr static const int32_t kNumShuffleTimes = 100;
+  constexpr static const int32_t kNums = 10;
+  constexpr static const double kUniquenessRate = .7;
+
+  template <typename T, bool returnsOptionalValues = false>
+  void testShuffle(const VectorPtr& input) {
+    DecodedVector decodedExpected;
+    decodedExpected.decode(*input.get());
+    exec::VectorReader<Array<T>> readerExpected(&decodedExpected);
+
+    auto actualVector =
+        evaluate<ArrayVector>("shuffle(C0)", makeRowVector({input}));
+    // Validate each row from the actual decoded ArrayVector is a permutation
+    // of the corresponding row from the expected decoded ArrayVector.
+    DecodedVector decodedActual;
+    decodedActual.decode(*actualVector.get());
+    exec::VectorReader<Array<T>> readerActual(&decodedActual);
+
+    for (auto i = 0; i < input->size(); i++) {
+      // Must materialize into std::vector, otherwise it will throw error
+      // because ArrayView doesn't support std::is_permutation() yet.
+      auto actualArray =
+          read<T, returnsOptionalValues>(readerActual, i).materialize();
```

Comment:
Since returnsOptionalValues is always true, this can be simplified to readerActual[i] and read function may be deleted.